{
    "active_lock_reason": null,
    "assignee": null,
    "assignees": [],
    "author_association": "CONTRIBUTOR",
    "body": "## \ud83d\udc1b Bug description\r\n\r\nMining cache tests fail on the latest SQLAlchemy version `1.4.0`.\r\n\r\nAll our CI tests will fail util this is resolved since in tox we install via `setup.py` and not via `requirements.txt`.\r\n\r\n## To reproduce\r\n\r\n```bash\r\npip install \"SQLAlchemy==1.4.0\"\r\npytest tests/test_database/test_mining_cache.py::TestCreateMiningCache\r\n```\r\nFail.\r\n\r\n```bash\r\npip install \"SQLAlchemy<1.4.0\"\r\npytest tests/test_database/test_mining_cache.py::TestCreateMiningCache\r\n```\r\nSuccess.\r\n\r\n## Details\r\nTraceback:\r\n```\r\n        cache_creator_instance = CreateMiningCache(\r\n            database_engine=fake_sqlalchemy_engine,\r\n            ee_models_library=ee_models_library,\r\n            target_table_name=\"mining_cache_temporary\",\r\n>           workers_per_model=2,\r\n        )\r\n\r\ntests/test_database/test_mining_cache.py:190: \r\n_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _\r\nsrc/bluesearch/database/mining_cache.py:282: in __init__\r\n    if not database_engine.dialect.has_table(database_engine, table_name):\r\nvenv/lib/python3.7/site-packages/sqlalchemy/dialects/sqlite/base.py:1999: in has_table\r\n    connection, \"table_info\", table_name, schema=schema\r\n_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _\r\n\r\nself = <sqlalchemy.dialects.sqlite.pysqlite.SQLiteDialect_pysqlite object at 0x126206750>\r\nconnection = Engine(sqlite:////private/var/folders/cz/v1lxwy5579502ksd81z14_t40000gp/T/pytest-of-sschmidt/pytest-6/db/cord19_test.db)\r\npragma = 'table_info', table_name = 'articles', schema = None\r\n\r\n    def _get_table_pragma(self, connection, pragma, table_name, schema=None):\r\n        quote = self.identifier_preparer.quote_identifier\r\n        if schema is not None:\r\n            statements = [\"PRAGMA %s.\" % quote(schema)]\r\n        else:\r\n            # because PRAGMA looks in all attached databases if no schema\r\n            # given, need to specify \"main\" schema, however since we want\r\n            # 'temp' tables in the same namespace as 'main', need to run\r\n            # the PRAGMA twice\r\n            statements = [\"PRAGMA main.\", \"PRAGMA temp.\"]\r\n    \r\n        qtable = quote(table_name)\r\n        for statement in statements:\r\n            statement = \"%s%s(%s)\" % (statement, pragma, qtable)\r\n>           cursor = connection.exec_driver_sql(statement)\r\nE           AttributeError: 'Engine' object has no attribute 'exec_driver_sql'\r\n\r\nvenv/lib/python3.7/site-packages/sqlalchemy/dialects/sqlite/base.py:2520: AttributeError\r\n```\r\n",
    "closed_at": "2021-03-16T15:35:38Z",
    "closed_by": {
        "avatar_url": "https://avatars.githubusercontent.com/u/8017698?v=4",
        "events_url": "https://api.github.com/users/Stannislav/events{/privacy}",
        "followers_url": "https://api.github.com/users/Stannislav/followers",
        "following_url": "https://api.github.com/users/Stannislav/following{/other_user}",
        "gists_url": "https://api.github.com/users/Stannislav/gists{/gist_id}",
        "gravatar_id": "",
        "html_url": "https://github.com/Stannislav",
        "id": 8017698,
        "login": "Stannislav",
        "node_id": "MDQ6VXNlcjgwMTc2OTg=",
        "organizations_url": "https://api.github.com/users/Stannislav/orgs",
        "received_events_url": "https://api.github.com/users/Stannislav/received_events",
        "repos_url": "https://api.github.com/users/Stannislav/repos",
        "site_admin": false,
        "starred_url": "https://api.github.com/users/Stannislav/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/Stannislav/subscriptions",
        "type": "User",
        "url": "https://api.github.com/users/Stannislav"
    },
    "comments": 3,
    "comments_content": [
        {
            "author_association": "CONTRIBUTOR",
            "body": "Here's a minimal example to reproduce this behaviour:\r\n```python\r\nimport os\r\nimport sqlite3\r\n\r\nimport sqlalchemy\r\n\r\n\r\ntable_name = \"test_table\"\r\ndb_name = \"example.db\"\r\n\r\n# Create a temporary SQLite database\r\ncon = sqlite3.connect(\"example.db\")\r\ncur = con.cursor()\r\ncur.execute(f\"CREATE TABLE {table_name} (name TEXT)\")\r\ncon.commit()\r\ncon.close()\r\n\r\n# Create an engine and check if the table exists\r\nengine = sqlalchemy.create_engine(f\"sqlite:///{db_name}\")\r\nprint(f\"Does the table {table_name} exist:\", engine.dialect.has_table(engine, table_name))\r\n\r\n# Remove the temporary SQLite database\r\nos.remove(db_name)\r\n```\r\n\r\nSeems like a bug in the latest `SQLAlchemy`. Maybe we should consider excluding the current version `1.4.0` in `setup.py`.",
            "created_at": "2021-03-16T13:22:54Z",
            "html_url": "https://github.com/BlueBrain/Search/issues/300#issuecomment-800253682",
            "id": 800253682,
            "issue_url": "https://api.github.com/repos/BlueBrain/Search/issues/300",
            "node_id": "MDEyOklzc3VlQ29tbWVudDgwMDI1MzY4Mg==",
            "performed_via_github_app": null,
            "reactions": {
                "+1": 0,
                "-1": 0,
                "confused": 0,
                "eyes": 0,
                "heart": 0,
                "hooray": 0,
                "laugh": 0,
                "rocket": 0,
                "total_count": 0,
                "url": "https://api.github.com/repos/BlueBrain/Search/issues/comments/800253682/reactions"
            },
            "updated_at": "2021-03-16T13:27:38Z",
            "url": "https://api.github.com/repos/BlueBrain/Search/issues/comments/800253682",
            "user": {
                "avatar_url": "https://avatars.githubusercontent.com/u/8017698?v=4",
                "events_url": "https://api.github.com/users/Stannislav/events{/privacy}",
                "followers_url": "https://api.github.com/users/Stannislav/followers",
                "following_url": "https://api.github.com/users/Stannislav/following{/other_user}",
                "gists_url": "https://api.github.com/users/Stannislav/gists{/gist_id}",
                "gravatar_id": "",
                "html_url": "https://github.com/Stannislav",
                "id": 8017698,
                "login": "Stannislav",
                "node_id": "MDQ6VXNlcjgwMTc2OTg=",
                "organizations_url": "https://api.github.com/users/Stannislav/orgs",
                "received_events_url": "https://api.github.com/users/Stannislav/received_events",
                "repos_url": "https://api.github.com/users/Stannislav/repos",
                "site_admin": false,
                "starred_url": "https://api.github.com/users/Stannislav/starred{/owner}{/repo}",
                "subscriptions_url": "https://api.github.com/users/Stannislav/subscriptions",
                "type": "User",
                "url": "https://api.github.com/users/Stannislav"
            }
        },
        {
            "author_association": "CONTRIBUTOR",
            "body": "Actually, I just found that after replacing\r\n```python\r\nengine.dialect.has_table(engine, table_name)\r\n```\r\nby\r\n```python\r\nengine.has_table(table_name)\r\n```\r\nthe exception goes away.\r\n\r\nStill seems like a bug in SQLAlchemy, but at least I think we can now create a fix for our repo.",
            "created_at": "2021-03-16T13:32:50Z",
            "html_url": "https://github.com/BlueBrain/Search/issues/300#issuecomment-800261406",
            "id": 800261406,
            "issue_url": "https://api.github.com/repos/BlueBrain/Search/issues/300",
            "node_id": "MDEyOklzc3VlQ29tbWVudDgwMDI2MTQwNg==",
            "performed_via_github_app": null,
            "reactions": {
                "+1": 0,
                "-1": 0,
                "confused": 0,
                "eyes": 0,
                "heart": 0,
                "hooray": 0,
                "laugh": 0,
                "rocket": 0,
                "total_count": 0,
                "url": "https://api.github.com/repos/BlueBrain/Search/issues/comments/800261406/reactions"
            },
            "updated_at": "2021-03-16T13:32:50Z",
            "url": "https://api.github.com/repos/BlueBrain/Search/issues/comments/800261406",
            "user": {
                "avatar_url": "https://avatars.githubusercontent.com/u/8017698?v=4",
                "events_url": "https://api.github.com/users/Stannislav/events{/privacy}",
                "followers_url": "https://api.github.com/users/Stannislav/followers",
                "following_url": "https://api.github.com/users/Stannislav/following{/other_user}",
                "gists_url": "https://api.github.com/users/Stannislav/gists{/gist_id}",
                "gravatar_id": "",
                "html_url": "https://github.com/Stannislav",
                "id": 8017698,
                "login": "Stannislav",
                "node_id": "MDQ6VXNlcjgwMTc2OTg=",
                "organizations_url": "https://api.github.com/users/Stannislav/orgs",
                "received_events_url": "https://api.github.com/users/Stannislav/received_events",
                "repos_url": "https://api.github.com/users/Stannislav/repos",
                "site_admin": false,
                "starred_url": "https://api.github.com/users/Stannislav/starred{/owner}{/repo}",
                "subscriptions_url": "https://api.github.com/users/Stannislav/subscriptions",
                "type": "User",
                "url": "https://api.github.com/users/Stannislav"
            }
        },
        {
            "author_association": "CONTRIBUTOR",
            "body": "I submitted an issue on the SQLAlchemy tracker: https://github.com/sqlalchemy/sqlalchemy/issues/6062",
            "created_at": "2021-03-16T13:55:48Z",
            "html_url": "https://github.com/BlueBrain/Search/issues/300#issuecomment-800277837",
            "id": 800277837,
            "issue_url": "https://api.github.com/repos/BlueBrain/Search/issues/300",
            "node_id": "MDEyOklzc3VlQ29tbWVudDgwMDI3NzgzNw==",
            "performed_via_github_app": null,
            "reactions": {
                "+1": 0,
                "-1": 0,
                "confused": 0,
                "eyes": 0,
                "heart": 0,
                "hooray": 0,
                "laugh": 0,
                "rocket": 0,
                "total_count": 0,
                "url": "https://api.github.com/repos/BlueBrain/Search/issues/comments/800277837/reactions"
            },
            "updated_at": "2021-03-16T13:55:48Z",
            "url": "https://api.github.com/repos/BlueBrain/Search/issues/comments/800277837",
            "user": {
                "avatar_url": "https://avatars.githubusercontent.com/u/8017698?v=4",
                "events_url": "https://api.github.com/users/Stannislav/events{/privacy}",
                "followers_url": "https://api.github.com/users/Stannislav/followers",
                "following_url": "https://api.github.com/users/Stannislav/following{/other_user}",
                "gists_url": "https://api.github.com/users/Stannislav/gists{/gist_id}",
                "gravatar_id": "",
                "html_url": "https://github.com/Stannislav",
                "id": 8017698,
                "login": "Stannislav",
                "node_id": "MDQ6VXNlcjgwMTc2OTg=",
                "organizations_url": "https://api.github.com/users/Stannislav/orgs",
                "received_events_url": "https://api.github.com/users/Stannislav/received_events",
                "repos_url": "https://api.github.com/users/Stannislav/repos",
                "site_admin": false,
                "starred_url": "https://api.github.com/users/Stannislav/starred{/owner}{/repo}",
                "subscriptions_url": "https://api.github.com/users/Stannislav/subscriptions",
                "type": "User",
                "url": "https://api.github.com/users/Stannislav"
            }
        }
    ],
    "comments_url": "https://api.github.com/repos/BlueBrain/Search/issues/300/comments",
    "created_at": "2021-03-16T10:27:35Z",
    "events_url": "https://api.github.com/repos/BlueBrain/Search/issues/300/events",
    "html_url": "https://github.com/BlueBrain/Search/issues/300",
    "id": 832646206,
    "labels": [
        {
            "color": "d73a4a",
            "default": false,
            "description": "Something isn't working",
            "id": 1938483675,
            "name": "\ud83d\udc1b bug fix",
            "node_id": "MDU6TGFiZWwxOTM4NDgzNjc1",
            "url": "https://api.github.com/repos/BlueBrain/Search/labels/%F0%9F%90%9B%20bug%20fix"
        }
    ],
    "labels_url": "https://api.github.com/repos/BlueBrain/Search/issues/300/labels{/name}",
    "locked": false,
    "milestone": null,
    "node_id": "MDU6SXNzdWU4MzI2NDYyMDY=",
    "number": 300,
    "performed_via_github_app": null,
    "reactions": {
        "+1": 0,
        "-1": 0,
        "confused": 0,
        "eyes": 0,
        "heart": 0,
        "hooray": 0,
        "laugh": 0,
        "rocket": 0,
        "total_count": 0,
        "url": "https://api.github.com/repos/BlueBrain/Search/issues/300/reactions"
    },
    "repository_url": "https://api.github.com/repos/BlueBrain/Search",
    "state": "closed",
    "state_reason": "completed",
    "timeline_url": "https://api.github.com/repos/BlueBrain/Search/issues/300/timeline",
    "title": "Tests fail on new SQLAlchemy",
    "updated_at": "2021-03-16T15:35:38Z",
    "url": "https://api.github.com/repos/BlueBrain/Search/issues/300",
    "user": {
        "avatar_url": "https://avatars.githubusercontent.com/u/8017698?v=4",
        "events_url": "https://api.github.com/users/Stannislav/events{/privacy}",
        "followers_url": "https://api.github.com/users/Stannislav/followers",
        "following_url": "https://api.github.com/users/Stannislav/following{/other_user}",
        "gists_url": "https://api.github.com/users/Stannislav/gists{/gist_id}",
        "gravatar_id": "",
        "html_url": "https://github.com/Stannislav",
        "id": 8017698,
        "login": "Stannislav",
        "node_id": "MDQ6VXNlcjgwMTc2OTg=",
        "organizations_url": "https://api.github.com/users/Stannislav/orgs",
        "received_events_url": "https://api.github.com/users/Stannislav/received_events",
        "repos_url": "https://api.github.com/users/Stannislav/repos",
        "site_admin": false,
        "starred_url": "https://api.github.com/users/Stannislav/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/Stannislav/subscriptions",
        "type": "User",
        "url": "https://api.github.com/users/Stannislav"
    }
}